### DISCUSSION TITLE
DB Migrations for Custom Modules

### DISCUSSION THREAD

**Discussion by TonyValenti** [oqtane/oqtane.framework#826] [Q&A]
Is there any documentation on how to do this?

**Comment by sbwalker** [comment]
See https://github.com/oqtane/oqtane.framework/issues/666

DB upgrades are complicated due to the multi-tenancy architecture which supports multiple databases in each installation. The SQL script based approach works fine. If you want to use Migrations it will require some familiarity with the multi-tenant concepts but it should be possible as @PoisnFang was successful

**Reply by PizzaConsole** [reply]
The way I was able to achieve it is a very hacked way, but I still prefer it over the SQL scripts.

**Reply by TonyValenti** [reply]
@PoisnFang - What would it take for you to be able to adjust the core so that it isn't such a hack?  Do you think you'd be able to make the adjustments necessary for it to be a robustly supported method?

**Reply by PizzaConsole** [reply]
That is a very good question. I will have to find some time to investigate.

**Reply by sbwalker** [reply]
@PoisnFang I am not sure if I would classify what you did as a "hack"... it uses the IInstallable interface which is the approproate framework extension point for this, and it utilizes the Tenant parameter.

**Reply by PizzaConsole** [reply]
@sbwalker Correct, applying the migrations is not as much of a hack. But actually making the migrations is. Check my comments, where I have to change a parent class and comment/uncomment code to make the migrations and then change it back afterwards:

```csharp
    // change to DbContext for ef core migration
    public class TodoContext : DBContextBase, IService
    {
        private readonly ITenantResolver _tenantResolver;
        private readonly IHttpContextAccessor _accessor;
        private readonly SiteState _siteState;

        public virtual DbSet<TodoList> TodoLists { get; set; }
        public virtual DbSet<TodoTask> TodoTasks { get; set; }
        public virtual DbSet<TodoUser> TodoUsers { get; set; }
        public virtual DbSet<Step> Steps { get; set; }

        // used for ef core migration
        private readonly string _connect = "Data Source=(LocalDb)\\MSSQLLocalDB;AttachDbFilename=|DataDirectory|\\Oqtane.mdf;Initial Catalog=Oqtane;Integrated Security=SSPI;";

        public Tenant InitializeTenant { get; set; }

        // comment out for ef core migration
        public TodoContext(ITenantResolver tenantResolver, IHttpContextAccessor accessor) : base(tenantResolver, accessor)
        {
            _tenantResolver = tenantResolver;
            _accessor = accessor;
        }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            // Uncomment to do ef core migration
            //InitializeTenant = new Tenant { DBConnectionString = _connect };

            if (InitializeTenant != null)
            {
                optionsBuilder.UseSqlServer(InitializeTenant.DBConnectionString.Replace("|DataDirectory|",
                    AppDomain.CurrentDomain.GetData("DataDirectory")?.ToString()));
            }
        }
```

**Reply by sbwalker** [reply]
I see... yes this is not optimal. This is a complicated area and would require some time/effort to determine a proper solution.

**Reply by skuflow** [reply]
@sbwalker  is there any movement on getting ef migrations to work properly?  I would have thought if you've built this on ef that migrations would be a fairly easy thing to implement.

**Reply by sbwalker** [reply]
@skuflow you are posting on a thread from 3 years ago... EF Migration support was added in 2021 ( the same time that support for multiple databases - SQL Server, SQLite, MySQL, and PostgreSQL was added). The best way to get started and see a working example is to use the Module Creator on the Develop page in a new install of the framework. It will ask you for some basic information and then scaffold all of the code for a fully functional module - including data access and migrations.

**Reply by sbwalker** [reply]
@skuflow and just to clarify, migrations were actually fairly challenging to implement due to the fact that Oqtane is multi-tenant
